processing/src/main/java/io/druid/segment/incremental/OakIncrementalIndex.java

```diff
@@ -0,0 +1,569 @@
+      @Override
+      public Boolean apply(Map.Entry<ByteBuffer, ByteBuffer> entry)
+      {
+        ByteBuffer serializedValue = entry.getValue();
+        BufferAggregator agg = getAggs()[aggIndex];
+        return agg.isNull(serializedValue, serializedValue.position() + aggsManager.aggOffsetInBuffer[aggIndex]);
+      }
+    };
+
+    OakTransformView<IncrementalIndexRow, Boolean> transformView = (OakTransformView<IncrementalIndexRow, Boolean>) oak.createTransformView(transformer);
+    return transformView.get(incrementalIndexRow);
+  }
+
+  @Override
+  public Iterable<IncrementalIndexRow> timeRangeIterable(
+          boolean descending, long timeStart, long timeEnd)
+  {
+    if (timeStart > timeEnd) {
+      return null;
+    }
+
+    IncrementalIndexRow from = new IncrementalIndexRow(timeStart, null, dimensionDescsList, IncrementalIndexRow.EMPTY_ROW_INDEX);
+    IncrementalIndexRow to = new IncrementalIndexRow(timeEnd + 1, null, dimensionDescsList, IncrementalIndexRow.EMPTY_ROW_INDEX);
+    OakMap subMap = oak.subMap(from, true, to, false);
+    if (descending == true) {
+      subMap = subMap.descendingMap();
+    }
+    CloseableIterator<IncrementalIndexRow> keysIterator = subMap.keysIterator();
+    return new Iterable<IncrementalIndexRow>() {
+      @Override
+      public Iterator<IncrementalIndexRow> iterator()
+      {
+        return Iterators.transform(
+          keysIterator,
+          key -> key
+        );
+      }
+    };
+  }
+
+  @Override
+  public Iterable<IncrementalIndexRow> keySet()
+  {
+    CloseableIterator<IncrementalIndexRow> keysIterator = oak.keysIterator();
+
+    return new Iterable<IncrementalIndexRow>() {
+      @Override
+      public Iterator<IncrementalIndexRow> iterator()
+      {
+        return Iterators.transform(
+            keysIterator,
+            key -> key
+        );
+      }
+    };
+  }
+
+  @Override
+  public boolean canAppendRow()
+  {
+    final boolean canAdd = size() < maxRowCount;
+    if (!canAdd) {
+      outOfRowsReason = StringUtils.format("Maximum number of rows [%d] reached", maxRowCount);
+    }
+    return canAdd;
+  }
+
+  private Integer addToOak(
+          InputRow row,
+          AtomicInteger numEntries,
+          IncrementalIndexRow incrementalIndexRow,
+          ThreadLocal<InputRow> rowContainer,
+          boolean skipMaxRowsInMemoryCheck
+  ) throws IndexSizeExceededException
+  {
+    OakComputer computer = new OakComputer(aggsManager, reportParseExceptions,
+            row, rowContainer);
+
+    if (isRollup()) {
+      incrementalIndexRow.setRowIndex(IncrementalIndexRow.EMPTY_ROW_INDEX);
+    } else {
+      incrementalIndexRow.setRowIndex(versionCounter.getAndIncrement());
+    }
+
+    if (numEntries.get() < maxRowCount || skipMaxRowsInMemoryCheck) {
+      oak.putIfAbsentComputeIfPresent(incrementalIndexRow, row, computer);
+
+      int currSize = oak.entries();
```

Comment:
Please add a comment explaining this loop. 
Also, what prevents it from being endless? Should we have some upper limit on the number of iterations?